modules/image-loading-optimization/module.php

```diff
@@ -0,0 +1,333 @@
+
+	/**
+	 * Get Module name.
+	 */
+	public function get_name() {
+		return 'image-loading-optimization';
+	}
+
+	/**
+	 * Constructor.
+	 */
+	public function __construct() {
+		parent::__construct();
+
+		$this->min_priority_img_pixels = 50000;
+		$this->omit_threshold = 3;
+
+		// Stop wp core logic.
+		add_action( 'init', [ $this, 'stop_core_fetchpriority_high_logic' ] );
+		add_filter( 'wp_lazy_loading_enabled', '__return_false' );
+
+		// Run optimization logic on header.
+		add_action( 'get_header', [ $this, 'set_buffer' ] );
+		// Ensure buffer is flushed (if any) before the content logic.
+		add_filter( 'the_content', [ $this, 'flush_header_buffer' ], 0 );
+
+		// Run optimization logic on content.
+		add_filter( 'wp_content_img_tag', [ $this, 'loading_optimization_image' ] );
+
+		// Run optimization logic on footer. Flushing of footer buffer will be handled by PHP script end default logic.
+		add_action( 'get_footer', [ $this, 'set_buffer' ] );
+	}
+
+	/**
+	 * Stop WordPress core fetchpriority logic by setting the wp_high_priority_element_flag flag to false.
+	 */
+	public function stop_core_fetchpriority_high_logic() {
+		// wp_high_priority_element_flag was only introduced in 6.3.0
+		if ( function_exists( 'wp_high_priority_element_flag' ) ) {
+			wp_high_priority_element_flag( false );
+		}
+	}
+
+	/**
+	 * Set buffer to handle header and footer content.
+	 */
+	public function set_buffer() {
+		ob_start( [ $this, 'handel_buffer_content' ] );
+	}
+
+	/**
+	 * This function ensure that buffer if any is flushed before the content is called.
+	 * This function behaves more like an action than a filter.
+	 *
+	 * @param string $content the content.
+	 * @return string We simply return the content from parameter.
+	 */
+	public function flush_header_buffer( $content ) {
+		$this->flush_buffer();
+		return $content;
+	}
+
+	/**
+	 * Flushes buffer.
+	 */
+	public function flush_buffer() {
+		ob_end_flush();
+	}
+
+	/**
+	 * Callback to handle image optimization logic on buffered content.
+	 *
+	 * @param string $buffer Buffered content.
+	 * @return string Content with optimized images.
+	 */
+	public function handel_buffer_content( $buffer ) {
+		return $this->filter_images( $buffer );
+	}
+
+	/**
+	 * Check for image in the content provided and apply optimization logic on them.
+	 *
+	 * @param string $content Content to be analyzed.
+	 * @return string Content with optimized images.
+	 */
+	private function filter_images( $content ) {
+		if ( ! preg_match_all( '/<img\s[^>]+>/', $content, $matches, PREG_SET_ORDER ) ) {
+			return $content;
+		}
+
+		// Iterate through the matches in order of occurrence as it is relevant for whether or not to lazy-load.
+		foreach ( $matches as $match ) {
+			$tag = $match[0];
+
+			// Optimize an image.
+			$optimized_image = $this->loading_optimization_image( $tag );
+
+			if ( $optimized_image !== $tag ) {
+				$content = str_replace( $tag, $optimized_image, $content );
+			}
+		}
```

Comment:
Instead of using `preg_match_all()` followed by `str_replace()`, I recommend instead a single call to `preg_replace_callback()`. This is much faster as it avoids re-searching throughout the content many times. This was an issue in core as well which I filed and fixed in https://core.trac.wordpress.org/ticket/58069